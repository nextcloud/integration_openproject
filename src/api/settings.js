/**
 * SPDX-FileCopyrightText: 2025 Jankari Tech Pvt. Ltd.
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import axios from '@nextcloud/axios'
import endpoints from './endpoints.js'

export function validateOPInstance(url) {
	return axios.post(endpoints.validateOPInstance, { url })
}

export function saveAdminConfig(configs) {
	return axios.put(endpoints.adminConfig, { values: configs })
}
