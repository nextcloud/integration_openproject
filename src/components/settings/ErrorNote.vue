<!--
  - SPDX-FileCopyrightText: 2025 Jankari Tech Pvt. Ltd.
  - SPDX-License-Identifier: AGPL-3.0-or-later
-->
<template>
	<NcNoteCard class="note-card" type="error">
		<p class="note-card--title">
			<b>{{ errorTitle }}</b>
		</p>
		<p class="note-card--error-description">
			<span v-if="errorMessage">{{ errorMessage }}. </span><a v-if="errorLink"
				class="link"
				:href="errorLink"
				target="_blank"
				:title="errorLinkLabel">{{ errorLinkLabel }}</a>
		</p>
	</NcNoteCard>
</template>

<script>
import { NcNoteCard } from '@nextcloud/vue'

export default {
	name: 'ErrorNote',
	components: {
		NcNoteCard,
	},
	props: {
		errorTitle: {
			type: String,
			required: true,
		},
		errorMessage: {
			type: String,
			default: '',
		},
		errorLink: {
			type: String,
			default: '',
		},
		errorLinkLabel: {
			type: String,
			default: 'Link',
		},
	},
}
</script>

<style lang="scss">
.note-card {
	max-width: 900px;
}

.link {
	color: #1a67a3 !important;
	font-style: normal;
}
</style>
