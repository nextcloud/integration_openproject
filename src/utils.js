/**
 * SPDX-FileCopyrightText: 2022-2025 Jankari Tech Pvt. Ltd.
 * SPDX-FileCopyrightText: 2021 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { showError, showSuccess } from '@nextcloud/dialogs'
import { translate as t } from '@nextcloud/l10n'

let mytimer = 0
export function delay(callback, ms) {
	return function() {
		const context = this
		const args = arguments
		clearTimeout(mytimer)
		mytimer = setTimeout(function() {
			callback.apply(context, args)
		}, ms || 0)
	}
}

export function checkOauthConnectionResult(oauthConnectionResult, oauthConnectionErrorMessage) {
	if (oauthConnectionResult === 'success') {
		showSuccess(t('integration_openproject', 'Successfully connected to OpenProject!'))
	} else if (oauthConnectionResult === 'error') {
		showError(
			t(
				'integration_openproject',
				'OAuth access token could not be obtained:',
			) + ' ' + oauthConnectionErrorMessage,
		)
	}
}

export const STATE = {
	OK: 'ok',
	ERROR: 'error',
	LOADING: 'loading',
	NO_TOKEN: 'no-token',
	CONNECTION_ERROR: 'connection-error',
	FAILED_FETCHING_WORKPACKAGES: 'failed-fetching-workpackages',
}

export const F_MODES = {
	VIEW: 0,
	EDIT: 1,
	DISABLE: 2,
	NEW: 3,
}

export const FORM = {
	SERVER: 0,
	AUTHORIZATION_METHOD: 1,
	AUTHORIZATION_SETTING: 2,
	OP_OAUTH: 3,
	NC_OAUTH: 4,
	GROUP_FOLDER: 5,
	APP_PASSWORD: 6,
}

export const WORKPACKAGES_SEARCH_ORIGIN = {
	PROJECT_TAB: 'project-tab',
	LINK_MULTIPLE_FILES_MODAL: 'link-multiple-files-modal',
}
export const USER_SETTINGS = {
	NAVIGATION_LINK_DESCRIPTION: t('integration_openproject', 'Displays a link to your OpenProject instance in the Nextcloud header.'),
	UNIFIED_SEARCH_DESCRIPTION: t('integration_openproject', 'Allows you to search OpenProject work packages via the universal search bar in Nextcloud.'),
}

export const NO_OPTION_TEXT_STATE = {
	START_TYPING: 0,
	SEARCHING: 1,
	RESULT: 2,
}

export const AUTH_METHOD = {
	OAUTH2: 'oauth2',
	OIDC: 'oidc',
}

export const AUTH_METHOD_LABEL = {
	OAUTH2: t('integration_openproject', 'Two-way OAuth 2.0 authorization code flow'),
	OIDC: t('integration_openproject', 'Single-Sign-On through OpenID Connect Identity Provider'),
}

export const SSO_PROVIDER_TYPE = {
	nextcloudHub: 'nextcloud_hub',
	external: 'external',
}

export const SSO_PROVIDER_LABEL = {
	nextcloudHub: 'Nextcloud Hub',
}
