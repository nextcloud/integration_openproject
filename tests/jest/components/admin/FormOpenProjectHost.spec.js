/* jshint esversion: 8 */

/**
 * SPDX-FileCopyrightText: 2025 Jankari Tech Pvt. Ltd.
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { createLocalVue, shallowMount } from '@vue/test-utils'
import { showError, showSuccess } from '@nextcloud/dialogs'
import flushPromises from 'flush-promises' // eslint-disable-line n/no-unpublished-import

import { F_MODES } from '../../../../src/utils.js'
import { validateOPInstance, saveAdminConfig } from '../../../../src/api/settings.js'
import FormOpenProjectHost from '../../../../src/components/admin/FormOpenProjectHost.vue'

// global mocks
global.t = (app, text) => text
// module mocks
jest.mock('@nextcloud/dialogs', () => ({
	getLanguage: jest.fn(() => ''),
	showError: jest.fn(),
	showSuccess: jest.fn(),
}))
jest.mock('../../../../src/api/settings.js', () => ({
	saveAdminConfig: jest.fn(() => ''),
	validateOPInstance: jest.fn(() => ({
		data: {
			result: true,
		},
	})),
}))

const localVue = createLocalVue()

const selectors = {
	formFieldValue: 'fieldvalue-stub',
	serverHostInput: '[data-test-id="openproject-server-host"]',
	saveFormButton: '[data-test-id="save-server-host"]',
	editFormButton: '[data-test-id="edit-server-host"]',
	cancelFormButton: '[data-test-id="cancel-server-host-edit"]',
}

describe('Component: FormOpenProjectHost', () => {
	afterEach(() => {
		jest.clearAllMocks()
	})

	describe('initial incomplete form', () => {
		let wrapper
		beforeEach(() => {
			wrapper = getWrapper()
		})

		it('should show the required form fields', async () => {
			expect(wrapper.find(selectors.serverHostInput).exists()).toBe(true)
			expect(wrapper.find(selectors.saveFormButton).exists()).toBe(true)
			expect(wrapper.find(selectors.formFieldValue).exists()).toBe(false)
			expect(wrapper.find(selectors.editFormButton).exists()).toBe(false)
			expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(false)
			expect(wrapper.html()).toMatchSnapshot()
		})
		it('should disable save button', async () => {
			expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe('true')
		})
		it('should enable save button when url is provided', async () => {
			const validUrl = 'http://example.openproject.test'
			const serverHostInput = wrapper.find(selectors.serverHostInput)

			serverHostInput.vm.$emit('input', validUrl)
			await flushPromises()

			expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe(undefined)
			expect(wrapper.vm.formDirty).toBe(true)
			expect(wrapper.vm.previousUrl).toBe('')
		})
		it('should disable save if url is empty', async () => {
			// set new url
			const serverHostInput = wrapper.find(selectors.serverHostInput)
			serverHostInput.vm.$emit('input', 'http://new.invalid.test')
			await flushPromises()
			expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe(undefined)
			expect(wrapper.vm.formDirty).toBe(true)

			// set empty url
			serverHostInput.vm.$emit('input', '')
			await flushPromises()
			expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe('true')
			expect(wrapper.vm.formDirty).toBe(false)
		})

		describe('valid url', () => {
			const validUrl = 'http://example.openproject.test'

			beforeEach(async () => {
				wrapper = getWrapper()
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', validUrl)
				await flushPromises()

				validateOPInstance.mockImplementation(() => ({
					data: {
						result: true,
					},
				}))
				saveAdminConfig.mockImplementation(() => Promise.resolve())
			})

			it('should save the url on submit', async () => {
				const saveFormButton = wrapper.find(selectors.saveFormButton)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe('')

				saveFormButton.vm.$emit('click')
				await flushPromises()

				expect(validateOPInstance).toHaveBeenCalledWith(validUrl)
				expect(saveAdminConfig).toHaveBeenCalledWith({ openproject_instance_url: validUrl })
				expect(showSuccess).toHaveBeenCalledTimes(1)
				expect(showError).toHaveBeenCalledTimes(0)
				expect(wrapper.vm.loading).toBe(false)
				expect(wrapper.vm.formMode).toBe(F_MODES.VIEW)
				expect(wrapper.vm.formDirty).toBe(false)
				expect(wrapper.vm.previousUrl).toBe(validUrl)
				expect(wrapper.emitted().formcomplete.length).toBe(1)
				expect(wrapper.emitted().formcomplete[0][0]).toBeInstanceOf(Function)

				expect(wrapper.find(selectors.editFormButton).exists()).toBe(true)
				expect(wrapper.find(selectors.formFieldValue).exists()).toBe(true)
				expect(wrapper.find(selectors.serverHostInput).exists()).toBe(false)
				expect(wrapper.find(selectors.saveFormButton).exists()).toBe(false)
				expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(false)
			})

			it('should show error on save failure', async () => {
				saveAdminConfig.mockImplementation(() => Promise.reject(new Error('Failure')))
				const saveFormButton = wrapper.find(selectors.saveFormButton)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)

				saveFormButton.vm.$emit('click')
				await flushPromises()

				expect(validateOPInstance).toHaveBeenCalledWith(validUrl)
				expect(saveAdminConfig).toHaveBeenCalledWith({ openproject_instance_url: validUrl })
				expect(showError).toHaveBeenCalledTimes(1)
				expect(showSuccess).toHaveBeenCalledTimes(0)
				expect(wrapper.vm.loading).toBe(false)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe('')
				expect(wrapper.emitted()).not.toHaveProperty('formcomplete')

				expect(wrapper.find(selectors.serverHostInput).exists()).toBe(true)
				expect(wrapper.find(selectors.saveFormButton).exists()).toBe(true)
				expect(wrapper.find(selectors.editFormButton).exists()).toBe(false)
				expect(wrapper.find(selectors.formFieldValue).exists()).toBe(false)
				expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(false)
			})
		})

		describe('invalid url', () => {
			beforeEach(async () => {
				wrapper = getWrapper()
			})

			it.each([
				{
					url: 'http://test.invalid.url',
					result: 'invalid',
					details: '',
					expectedMessage: 'URL is invalid',
					expectedDetails: 'The URL should have the form "https://openproject.org"',
				},
				{
					url: 'http://not.found.test',
					result: 'not_valid_body',
					details: '<body>the complete body of the return</body>',
					expectedMessage: 'There is no valid OpenProject instance listening at that URL, please check the Nextcloud logs',
					expectedDetails: '',
				},
				{
					url: 'http://not.found.test',
					result: 'client_exception',
					details: '404 Not Found',
					expectedMessage: 'There is no valid OpenProject instance listening at that URL, please check the Nextcloud logs',
					expectedDetails: 'Response: "404 Not Found"',
				},
				{
					url: 'http://example.openproject.test',
					result: 'server_exception',
					details: '503 Service Unavailable',
					expectedMessage: 'Server replied with an error message, please check the Nextcloud logs',
					expectedDetails: '503 Service Unavailable',
				},
				{
					url: 'http://local.test',
					result: 'local_remote_servers_not_allowed',
					details: '',
					expectedMessage: 'Accessing OpenProject servers with local addresses is not allowed.',
					expectedDetails: 'To be able to use an OpenProject server with a local address, enable the `allow_local_remote_servers` setting. {htmlLink}.',
				},
				{
					url: 'http://redirect.openproject.test',
					result: 'redirected',
					details: 'https://redirected.url',
					expectedMessage: 'The given URL redirects to \'{location}\'. Please do not use a URL that leads to a redirect.',
					expectedDetails: '',
				},
				{
					url: 'http://example.openproject.test',
					result: 'unexpected_error',
					details: '',
					expectedMessage: 'Could not connect to the given URL, please check the Nextcloud logs',
					expectedDetails: '',
				},
				{
					url: 'http://example.openproject.test',
					result: 'network_error',
					details: '',
					expectedMessage: 'Could not connect to the given URL, please check the Nextcloud logs',
					expectedDetails: '',
				},
				{
					url: 'http://example.openproject.test',
					result: 'request_exception',
					details: '',
					expectedMessage: 'Could not connect to the given URL, please check the Nextcloud logs',
					expectedDetails: '',
				},
			])('$result: should show error if url is invalid', async ({ url, result, details, expectedMessage, expectedDetails }) => {
				validateOPInstance.mockImplementation(() => ({
					data: {
						result,
						details,
					},
				}))
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', url)
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe('')

				const saveFormButton = wrapper.find(selectors.saveFormButton)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)

				saveFormButton.vm.$emit('click')
				await flushPromises()

				expect(validateOPInstance).toHaveBeenCalledWith(url)
				expect(saveAdminConfig).not.toHaveBeenCalled()
				expect(wrapper.vm.errorMessage).toBe(expectedMessage)
				expect(wrapper.vm.errorDetails).toBe(expectedDetails)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe('')
				expect(wrapper.emitted()).not.toHaveProperty('formcomplete')

				expect(wrapper.find(selectors.serverHostInput).exists()).toBe(true)
				expect(wrapper.find(selectors.saveFormButton).exists()).toBe(true)
				expect(wrapper.find(selectors.editFormButton).exists()).toBe(false)
				expect(wrapper.find(selectors.formFieldValue).exists()).toBe(false)
				expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(false)
				expect(wrapper.html()).toMatchSnapshot()
			})
		})
	})

	describe('completed form', () => {
		const validUrl = 'http://example.openproject.test'
		let wrapper

		beforeEach(() => {
			wrapper = getWrapper({}, { openprojectUrl: validUrl })
		})

		it('should show text field value with edit button', () => {
			expect(wrapper.find(selectors.formFieldValue).exists()).toBe(true)
			expect(wrapper.find(selectors.editFormButton).exists()).toBe(true)
			expect(wrapper.vm.formMode).toBe(F_MODES.VIEW)
			expect(wrapper.find(selectors.serverHostInput).exists()).toBe(false)
			expect(wrapper.find(selectors.saveFormButton).exists()).toBe(false)
			expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(false)

			expect(wrapper.emitted().formcomplete[0][0]).toBeInstanceOf(Function)
			expect(wrapper.vm.serverUrl).toBe(validUrl)
			expect(wrapper.vm.previousUrl).toBe(validUrl)
			expect(wrapper.html()).toMatchSnapshot()
		})

		describe('edit mode', () => {
			beforeEach(async () => {
				wrapper = getWrapper({}, { openprojectUrl: validUrl })
				const editButton = wrapper.find(selectors.editFormButton)
				editButton.vm.$emit('click')
				await flushPromises()
			})

			it('should show input form fields', async () => {
				expect(wrapper.find(selectors.formFieldValue).exists()).toBe(false)
				expect(wrapper.find(selectors.editFormButton).exists()).toBe(false)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)
				expect(wrapper.find(selectors.serverHostInput).exists()).toBe(true)
				expect(wrapper.find(selectors.saveFormButton).exists()).toBe(true)
				expect(wrapper.find(selectors.cancelFormButton).exists()).toBe(true)
				expect(wrapper.html()).toMatchSnapshot()
			})
			it('should have disabled save button and enabled cancel button', async () => {
				expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe('true')
				expect(wrapper.find(selectors.cancelFormButton).attributes().disabled).toBe(undefined)
			})
			it('should show form in view mode on cancel', async () => {
				const cancelButton = wrapper.find(selectors.cancelFormButton)
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)

				// set new url
				const newUrl = 'http://new.openproject.test'
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', newUrl)
				await flushPromises()
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe(validUrl)

				// save button should be enabled
				expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe(undefined)
				expect(wrapper.find(selectors.cancelFormButton).attributes().disabled).toBe(undefined)

				await cancelButton.vm.$emit('click')
				await flushPromises()

				expect(wrapper.vm.formMode).toBe(F_MODES.VIEW)
				expect(wrapper.vm.serverUrl).toBe(validUrl)
				expect(wrapper.vm.formDirty).toBe(false)
				expect(wrapper.vm.previousUrl).toBe(validUrl)
			})
			it('should reset error messages on cancel', async () => {
				validateOPInstance.mockImplementation(() => ({
					data: {
						result: 'invalid',
						details: '',
					},
				}))
				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)

				// set new url
				const newUrl = 'http://new.invalid.test'
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', newUrl)
				await flushPromises()
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe(validUrl)

				const saveButton = wrapper.find(selectors.saveFormButton)
				await saveButton.vm.$emit('click')
				await flushPromises()
				expect(wrapper.vm.errorMessage).toBe('URL is invalid')
				expect(wrapper.vm.errorDetails).toBe('The URL should have the form "https://openproject.org"')

				const cancelButton = wrapper.find(selectors.cancelFormButton)
				await cancelButton.vm.$emit('click')
				await flushPromises()

				expect(wrapper.vm.formMode).toBe(F_MODES.VIEW)
				expect(wrapper.vm.serverUrl).toBe(validUrl)
				expect(wrapper.vm.formDirty).toBe(false)
				expect(wrapper.vm.previousUrl).toBe(validUrl)
				expect(wrapper.vm.errorMessage).toBe('')
				expect(wrapper.vm.errorDetails).toBe('')
			})
			it('should disable save if url is empty', async () => {
				// set new url
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', 'http://new.invalid.test')
				await flushPromises()
				expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe(undefined)
				expect(wrapper.vm.formDirty).toBe(true)

				// set empty url
				serverHostInput.vm.$emit('input', '')
				await flushPromises()
				expect(wrapper.find(selectors.saveFormButton).attributes().disabled).toBe('true')
				expect(wrapper.vm.formDirty).toBe(false)
			})
			it('should show form in view mode on save', async () => {
				validateOPInstance.mockImplementation(() => ({
					data: {
						result: true,
					},
				}))
				saveAdminConfig.mockImplementation(() => Promise.resolve())

				expect(wrapper.vm.formMode).toBe(F_MODES.EDIT)
				// set new url
				const newUrl = 'http://new.openproject.test'
				const serverHostInput = wrapper.find(selectors.serverHostInput)
				serverHostInput.vm.$emit('input', newUrl)
				await flushPromises()
				expect(wrapper.vm.formDirty).toBe(true)
				expect(wrapper.vm.previousUrl).toBe(validUrl)

				const saveButton = wrapper.find(selectors.saveFormButton)
				await saveButton.vm.$emit('click')
				await flushPromises()

				expect(validateOPInstance).toHaveBeenCalledWith(newUrl)
				expect(saveAdminConfig).toHaveBeenCalledWith({ openproject_instance_url: newUrl })
				expect(showSuccess).toHaveBeenCalledTimes(1)
				expect(showError).toHaveBeenCalledTimes(0)
				expect(wrapper.vm.formMode).toBe(F_MODES.VIEW)
				expect(wrapper.vm.formDirty).toBe(false)
				expect(wrapper.vm.serverUrl).toBe(newUrl)
				expect(wrapper.vm.previousUrl).toBe(newUrl)
				expect(wrapper.emitted().formcomplete.length).toBe(1)
			})
		})
	})
})

function getWrapper(data = {}, props = {}) {
	return shallowMount(FormOpenProjectHost, {
		localVue,
		mocks: {
			t: (app, msg) => msg,
		},
		propsData: props,
		data() {
			return data
		},
	})
}
