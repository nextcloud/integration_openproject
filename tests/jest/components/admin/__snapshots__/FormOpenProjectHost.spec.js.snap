// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FormOpenProjectHost completed form edit mode should show input form fields 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server" iscomplete="true"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://example.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <ncbutton-stub alignment="center" type="secondary" nativetype="button" data-test-id="cancel-server-host-edit" class="mr-2">
        Cancel
      </ncbutton-stub>
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host" disabled="true">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost completed form should show text field value with edit button 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server" iscomplete="true"></formheading-stub>
  <div class="openproject-host">
    <fieldvalue-stub title="OpenProject host" value="http://example.openproject.test" isrequired="true" class="pb-1"></fieldvalue-stub>
    <div class="form-actions">
      <ncbutton-stub alignment="center" type="secondary" nativetype="button" data-test-id="edit-server-host">
        Edit server information
      </ncbutton-stub>
      <!---->
      <!---->
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url client_exception: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://not.found.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="There is no valid OpenProject instance listening at that URL, please check the Nextcloud logs" errormessagedetails="Response: &quot;404 Not Found&quot;" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url invalid: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://test.invalid.url" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="URL is invalid" errormessagedetails="The URL should have the form &quot;https://openproject.org&quot;" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url local_remote_servers_not_allowed: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://local.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="Accessing OpenProject servers with local addresses is not allowed." errormessagedetails="To be able to use an OpenProject server with a local address, enable the \`allow_local_remote_servers\` setting. {htmlLink}." isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url network_error: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://example.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="Could not connect to the given URL, please check the Nextcloud logs" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url not_valid_body: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://not.found.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="There is no valid OpenProject instance listening at that URL, please check the Nextcloud logs" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url redirected: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://redirect.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="The given URL redirects to '{location}'. Please do not use a URL that leads to a redirect." errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url request_exception: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://example.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="Could not connect to the given URL, please check the Nextcloud logs" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url server_exception: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://example.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="Server replied with an error message, please check the Nextcloud logs" errormessagedetails="503 Service Unavailable" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form invalid url unexpected_error: should show error if url is invalid 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="http://example.openproject.test" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="Could not connect to the given URL, please check the Nextcloud logs" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;

exports[`Component: FormOpenProjectHost initial incomplete form should show the required form fields 1`] = `
"<div id="server-host">
  <formheading-stub index="1" title="OpenProject server"></formheading-stub>
  <div class="openproject-host">
    <textinput-stub value="" type="text" placeholder="https://www.my-openproject.com" label="OpenProject host" hinttext="Please introduce your OpenProject hostname" errormessage="" errormessagedetails="" isrequired="true" data-test-id="openproject-server-host" class="pb-1"></textinput-stub>
    <div class="form-actions">
      <!---->
      <!---->
      <ncbutton-stub alignment="center" disabled="true" type="primary" nativetype="button" data-test-id="save-server-host">
        Save
      </ncbutton-stub>
    </div>
  </div>
</div>"
`;
